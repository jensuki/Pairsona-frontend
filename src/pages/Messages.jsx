import { useState, useEffect, useContext, useRef } from 'react';
import { useParams } from 'react-router-dom';
import Api from '../helpers/api';
import UserContext from '../context/UserContext';
import '../styles/Messages.css';

const Messages = () => {
    const { username } = useParams();
    const { currUser, loading, checkNotifications } = useContext(UserContext);
    const [messages, setMessages] = useState([]);
    const [newMessage, setNewMessage] = useState('');
    const [error, setError] = useState(null);
    const messageHistoryRef = useRef(null); // ref for msg history container

    // scroll to recent msg
    const scrollToLatest = () => {
        if (messageHistoryRef.current) {
            messageHistoryRef.current.scrollTop = messageHistoryRef.current.scrollHeight;
        }
    }

    // fetch message history on mount
    useEffect(() => {
        const fetchMessages = async () => {
            try {
                const res = await Api.getMessages(username);
                setMessages(res);

                // mark all messages as read
                for (let msg of res) {
                    if (!msg.is_read && msg.recipient_id === currUser.id) {
                        await Api.markMessageAsRead(msg.id);
                    }
                }
                // refresh notifications after marking messages as read
                checkNotifications();
                scrollToLatest();
            } catch (err) {
                setError(err.message || 'Error fetching messages');
            }
        };

        fetchMessages();
    }, [username, currUser, checkNotifications]);

    // scroll to bottom whenever messages are updated
    useEffect(() => {
        scrollToLatest();
    }, [messages]);

    // handle sending a new message
    const handleSendMessage = async (e) => {
        e.preventDefault();
        setError(null); // Clear previous errors

        try {
            const res = await Api.sendMessage(username, newMessage);

            if (!res) {
                return; // let global token expiration handler manage redirection/toast
            }

            setMessages([...messages, res.message]); // append new msg
            setNewMessage('');
        } catch (err) {
            console.error('Error sending message:', err);
            setError(err.message || 'Error sending message');
        }
    };

    if (loading) return <p>Loading...</p>;

    return (
        <div className="Messages">
            <h2>Chat with {username}</h2>
            <div className="message-history" ref={messageHistoryRef}>
                {/* render messages with separate 'sent' and 'received' classes */}
                {messages.map((msg) => (
                    <div
                        key={msg.id}
                        className={`message ${currUser.id && msg.sender_id === currUser.id ? 'sent' : 'received'
                            }`}>
                        <p>{msg.content}</p>
                        <small>{new Date(msg.created_at).toLocaleString()}</small>
                    </div>
                ))}
            </div>
            <form onSubmit={handleSendMessage}>
                <input
                    type="text"
                    value={newMessage}
                    onChange={(e) => setNewMessage(e.target.value)}
                    placeholder="Type your message..."
                    required
                />
                <button type="submit">Send</button>
                {error && <p className="error" role="alert">{error}</p>}
            </form>
        </div>
    );
};

export default Messages;
