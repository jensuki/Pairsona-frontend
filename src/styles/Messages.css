.Messages {
    max-width: var(--max-width-sm);
}

.Messages h2 {
    margin-top: 0;
}

.message-history {
    max-height: 300px;
    overflow-y: auto;
    scroll-behavior: smooth;
    margin-bottom: 1rem;
    padding: 0 1rem;
}

.message {
    padding: .1rem 1rem .7rem 1rem;
    margin: .5rem 0;
    border-radius: var(--border-radius-md);
    position: relative;
}

.sent {
    background: linear-gradient(to left, #d1e7dd 60%, transparent);
    text-align: right;
    border-radius: var(--border-radius-md);
}

.sent::before {
    content: '';
    position: absolute;
    top: 52px;
    right: -15px;
    border-width: 10px 0px 10px 20px;
    border-style: solid;
    border-color: transparent transparent transparent #d1e7dd;
}

.received {
    background: linear-gradient(to right, #f8d7da 60%, transparent);
    text-align: left;
    border-radius: var(--border-radius-md);
}

.received::before {
    content: '';
    position: absolute;
    bottom: 15px;
    left: -15px;
    border-width: 10px 20px 10px 0px;
    border-style: solid;
    border-color: transparent #f8d7da transparent transparent;
}

.Messages form {
    display: flex;
    gap: .5rem;
    max-width: calc(var(--max-width-sm) - 25px);
    margin: 0 auto;
}

.Messages input[type="text"] {
    flex-grow: 1;
    padding: .5rem;
    border: 1px solid #ccc;
    border-radius: var(--border-radius-sm);
}

.Messages button {
    padding: .5rem .8rem;
}