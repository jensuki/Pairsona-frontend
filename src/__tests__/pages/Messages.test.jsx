import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../test_common/setupTests';
import Messages from '../../pages/Messages';
import Api from '../../helpers/api';
import { describe, expect, test, vi, beforeEach } from 'vitest';
import { demoUser } from '../../test_common/mockData';

// mock api methods
vi.mock('../../helpers/api', () => ({

    default: {
        getMessages: vi.fn(),
        markMessageAsRead: vi.fn(),
        sendMessage: vi.fn()
    }
}))

// mock userParams
vi.mock('react-router-dom', async () => {
    const actual = await vi.importActual('react-router-dom');
    return {
        ...actual, // import actual module
        useParams: () => ({ username: 'user3' }), // mock username to return 'user3'
    };
});

// helper
const renderMessages = () => {
    return render(
        <MemoryRouter>
            <UserProvider currUser={demoUser}>
                <Messages />
            </UserProvider>
        </MemoryRouter>
    )
}

describe('Messages component', () => {
    beforeEach(() => {
        vi.clearAllMocks();

        // mock api repsonse to return predefined array of msgs
        Api.getMessages.mockResolvedValue([
            {
                id: 1,
                sender_id: 2,
                recipient_id: 1, // demouser receives
                content: 'Hello!',
                created_at: '2023-01-01T12:00:00Z',
                is_read: false, // unread
            },
            {
                id: 2,
                sender_id: 1, // demouser sends
                recipient_id: 2,
                content: 'Hi there!',
                created_at: '2023-01-01T12:01:00Z',
                is_read: true, // read
            },
        ]);
    });

    test('renders without crashing and displays username from useParams', async () => {
        renderMessages();

        // wait for 'user3' to be displayed in dom
        await waitFor(() =>
            expect(screen.getByText(/Chat with user3/i)).toBeInTheDocument()
        );
        // verify mock msgs are rendered
        expect(screen.getByText('Hello!')).toBeInTheDocument()
        expect(screen.getByText('Hi there!')).toBeInTheDocument();
    });

    test('matches snapshot', async () => {
        const { asFragment } = renderMessages();

        await waitFor(() => {
            expect(asFragment()).toMatchSnapshot();
        })

    });

    test('marks unread messages as read on render', async () => {
        renderMessages();

        await waitFor(() => {
            expect(Api.markMessageAsRead).toHaveBeenCalledWith(1); // id of the unread msg
        })
    })
    test('sends a new message', async () => {
        // mock a new message being sent
        Api.sendMessage.mockResolvedValue({
            message: {
                id: 3,
                sender_id: 1,
                recipient_id: 2,
                content: 'New message!',
                created_at: '2023-01-01T12:02:00Z',
            },
        });

        renderMessages();

        // simulate typing the new msg
        const input = screen.getByPlaceholderText('Type your message...');
        fireEvent.change(input, { target: { value: 'New message!' } });

        // click send
        const sendBtn = screen.getByText('Send');
        fireEvent.click(sendBtn);

        // wait for new msg to appear
        await waitFor(() => {
            expect(screen.getByText('New message!')).toBeInTheDocument();
        })

        // ensure api was called
        expect(Api.sendMessage).toHaveBeenCalledWith('user3', 'New message!')
    })
});