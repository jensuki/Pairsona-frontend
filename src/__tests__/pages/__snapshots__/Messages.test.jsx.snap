// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Messages component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Messages"
  >
    <h2>
      Chat with user3
    </h2>
    <div
      class="message-history"
    />
    <form>
      <input
        placeholder="Type your message..."
        required=""
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Send
      </button>
    </form>
  </div>
</DocumentFragment>
`;
